crates/bifrost/src/bifrost.rs

```diff
@@ -99,4 +146,216 @@ impl BifrostInner {
+            let lsn = bifrost
+                .append(
+                    LogId::from(0),
+                    DataRecord::default(),
+                    AppendAttributes::default(),
+                )
+                .await?;
+            info!(%lsn, "Appended record to log");
+            assert_eq!(Lsn(i), lsn);
+            max_lsn = lsn;
+        }
+
+        // Append to a log that doesn't exist.
+        let invalid_log = LogId::from(num_partitions + 1);
+        let resp = bifrost
+            .append(
+                invalid_log,
+                DataRecord::default(),
+                AppendAttributes::default(),
+            )
+            .await;
+
+        assert_that!(resp, pat!(Err(pat!(Error::UnknownLogId(eq(invalid_log))))));
+
+        // use a cloned bifrost.
+        let mut cloned_bifrost = bifrost.clone();
+        for _ in 1..=5 {
+            // Append a record to memory
+            let lsn = cloned_bifrost
+                .append(
+                    LogId::from(0),
+                    DataRecord::default(),
+                    AppendAttributes::default(),
+                )
+                .await?;
+            info!(%lsn, "Appended record to log");
+            assert_eq!(max_lsn + Lsn(1), lsn);
+            max_lsn = lsn;
+        }
+
+        // Ensure original clone writes to the same underlying loglet.
+        let lsn = clean_bifrost_clone
+            .append(
+                LogId::from(0),
+                DataRecord::default(),
+                AppendAttributes::default(),
+            )
+            .await?;
+        assert_eq!(max_lsn + Lsn(1), lsn);
+        max_lsn = lsn;
+
+        // Writes to a another log doesn't impact existing
+        let lsn = bifrost
+            .append(
+                LogId::from(3),
+                DataRecord::default(),
+                AppendAttributes::default(),
+            )
+            .await?;
+        assert_eq!(Lsn(1), lsn);
+
+        let lsn = bifrost
+            .append(
+                LogId::from(0),
+                DataRecord::default(),
+                AppendAttributes::default(),
+            )
+            .await?;
+        assert_eq!(max_lsn + Lsn(1), lsn);
+
+        // Initiate shutdown
+        shutdown_signal.drain().await;
+
+        assert!(svc_handle.is_finished());
+
+        // appends cannot succeed after shutdown
+        let res = bifrost
+            .append(
+                LogId::from(0),
+                DataRecord::default(),
+                AppendAttributes::default(),
+            )
+            .await;
+        assert!(matches!(res, Err(Error::Shutdown)));
+        // Validate the watchdog has called the provider::start() function.
+        assert!(logs_contain("Starting in-memory loglet provider"));
+        assert!(logs_contain("Shutting down in-memory loglet provider"));
+        assert!(logs_contain("Bifrost watchdog shutdown complete"));
+        Ok(())
+    }
+
+    #[test(tokio::test(start_paused = true))]
```

Comment:
I didn't know about `start_paused = true` and then using `tokio::time::Instant`. This is really cool :-)